### PR TITLE
feat: Add Azure Key Vault Variable Provider

### PR DESCRIPTION
**Pull request by ThorstenHans**
This PR adds Azure Key Vault as Variable provider. 

Users can provide a runtime config using necessary information for Client Credential Flow:

```toml
[[config_provider]]
type = "azure_key_vault"
vault_url = "https://some.vault.azure.net/"
client_id = "11111111-1111-1111-1111-111111111111"
client_secret = "11111111-1111-1111-1111-111111111111"
tenant_id = "11111111-1111-1111-1111-111111111111"
# authority_host = "AzurePublicCloud"
```

If not specified, `authority_host` will default to `AzurePublicCloud`.

Unfortunately, I wasn't able to run `make lint` and `make test` on my machine. Both commands ran into errors shown below. 

Am I missing something on my machine? 

### Output from `make lint`

```bash
    Checking spin-expressions v2.5.0-pre0 (/Users/thorsten/dev/thorstenhans/spin/crates/expressions)
error: you are explicitly cloning with `.map()`
  --> crates/plugins/src/badger/store.rs:89:26
   |
89 |             badgered_to: to.iter().map(|v| <semver::Version>::clone(v)).collect(),
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider calling the dedicated `cloned` method: `to.iter().cloned()`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_clone
   = note: `-D clippy::map-clone` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::map_clone)]`

    Checking spin-outbound-networking v2.5.0-pre0 (/Users/thorsten/dev/thorstenhans/spin/crates/outbound-networking)
    Checking spin-doctor v2.5.0-pre0 (/Users/thorsten/dev/thorstenhans/spin/crates/doctor)
    Checking spin-build v2.5.0-pre0 (/Users/thorsten/dev/thorstenhans/spin/crates/build)
    Checking spin-templates v2.5.0-pre0 (/Users/thorsten/dev/thorstenhans/spin/crates/templates)
error: could not compile `spin-plugins` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `spin-plugins` (lib test) due to 1 previous error
make: *** [lint] Error 101
```

### Output from `make test`

```bash
Checking spin-key-value v2.5.0-pre0 (/Users/thorsten/dev/thorstenhans/spin/crates/key-value)
error: you are explicitly cloning with `.map()`
  --> crates/plugins/src/badger/store.rs:89:26
   |
89 |             badgered_to: to.iter().map(|v| <semver::Version>::clone(v)).collect(),
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider calling the dedicated `cloned` method: `to.iter().cloned()`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_clone
   = note: `-D clippy::map-clone` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::map_clone)]`

    Checking outbound-http v2.5.0-pre0 (/Users/thorsten/dev/thorstenhans/spin/crates/outbound-http)
error: could not compile `spin-plugins` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `spin-plugins` (lib test) due to 1 previous error
error: field `0` is never read
   --> crates/templates/src/template.rs:649:21
    |
649 |     struct TempFile(tempfile::TempDir, PathBuf);
    |            -------- ^^^^^^^^^^^^^^^^^
    |            |
    |            field in this struct
    |
    = note: `-D dead-code` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(dead_code)]`
help: consider changing the field to be of unit type to suppress this warning while preserving the field numbering, or remove the field
    |
649 |     struct TempFile((), PathBuf);
    |                     ~~

error: could not compile `spin-templates` (lib test) due to 1 previous error
make: *** [lint] Error 101
```

### CODE REVIEW

**Comment by rylev** [comment]
If you rebase against the latest `main` branch, the errors you are seeing should be fixed. They were address [here](https://github.com/fermyon/spin/pull/2474).

File: crates/variables/src/provider/azkv.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+            self.tenant_id.to_string(),
+            self.client_id.to_string(),
+            self.client_secret.to_string(),
+        );
+
+        let secret_client = SecretClient::new(&self.vault_url, Arc::new(credential))?;
+        match secret_client.get(key.as_str()).await {
+            Ok(secret) => Ok(Some(secret.value)),
+            Err(err) => return Err(err).context("Failed to read variable from Azure Key Vault"),
+        }
```
Comment (rylev) [inline crates/variables/src/provider/azkv.rs:55]:
```suggestion
        let secret = secret_client.get(key.as_str()).await.context("Failed to read variable from Azure Key Vault")?;
        Ok(Some(secret.value))
         
```

File: crates/variables/src/provider/azkv.rs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    }
+}
+
+impl From<AzureAuthorityHost> for Url {
+    fn from(value: AzureAuthorityHost) -> Self {
+        let url = match value {
+            AzureAuthorityHost::AzureChina => "https://login.chinacloudapi.cn/",
+            AzureAuthorityHost::AzureGovernment => "https://login.microsoftonline.us/",
+            AzureAuthorityHost::AzureGermany => "https://login.microsoftonline.de/",
+            _ => "https://login.microsoftonline.com/",
```
Comment (rylev) [inline crates/variables/src/provider/azkv.rs:79]:
```suggestion
            AzureAuthorityHost::AzurePublicCloud => "https://login.microsoftonline.com/",
```

This will prevent potential subtle bugs if a new variant of `AzureAuthorityHost` is added.

File: crates/variables/src/provider/azkv.rs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            "azuregermany" | "germany" => AzureAuthorityHost::AzureGermany,
+            "AzureGovernment" | "gov" => AzureAuthorityHost::AzureGovernment,
+            _ => AzureAuthorityHost::AzurePublicCloud,
+        }
+    }
+}
+
+impl From<String> for AzureAuthorityHost {
+    fn from(value: String) -> Self {
+        AzureAuthorityHost::from(value.as_str())
```
Comment (rylev) [inline crates/variables/src/provider/azkv.rs:97]:
```suggestion
        value.as_str().into()
```

File: crates/variables/src/provider/azkv.rs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        Url::parse(url).unwrap()
+    }
+}
+impl From<&str> for AzureAuthorityHost {
+    fn from(value: &str) -> Self {
+        match value.to_lowercase().as_str() {
+            "azurechina" | "china" => AzureAuthorityHost::AzureChina,
+            "azuregermany" | "germany" => AzureAuthorityHost::AzureGermany,
+            "AzureGovernment" | "gov" => AzureAuthorityHost::AzureGovernment,
+            _ => AzureAuthorityHost::AzurePublicCloud,
```
Comment (rylev) [inline crates/variables/src/provider/azkv.rs:90]: Are we sure we want to accept any string? 

You want want to implement [`FromStr`](https://doc.rust-lang.org/stable/std/str/trait.FromStr.html) instead and then you can use the `parse` function on `&str`.

**Review by rylev — APPROVED** [review]
Looks great overall! I did not actually run the code, but it looks pretty straight forward to me. I think we can merge this. The only points I had were nitpicks.

File: crates/variables/src/provider.rs [hunk 5]
```diff
@@ -1,2 +1,3 @@
+pub mod azkv;
```
Comment (itowlson) [inline crates/variables/src/provider.rs:1]: `azure_kv` or `azure_key_vault` might be a more self-explanatory name.

File: crates/variables/src/provider/azkv.rs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        };
+        Url::parse(url).unwrap()
+    }
+}
+impl From<&str> for AzureAuthorityHost {
+    fn from(value: &str) -> Self {
+        match value.to_lowercase().as_str() {
+            "azurechina" | "china" => AzureAuthorityHost::AzureChina,
+            "azuregermany" | "germany" => AzureAuthorityHost::AzureGermany,
+            "AzureGovernment" | "gov" => AzureAuthorityHost::AzureGovernment,
```
Comment (itowlson) [inline crates/variables/src/provider/azkv.rs:89]: This is matching a lowercased string against a mixed-case one - is that intentional?

Comment (ThorstenHans) [reply crates/variables/src/provider/azkv.rs]: fixed with 0a3e792

File: crates/variables/src/provider.rs (hunk 5 above)
Comment (ThorstenHans) [reply crates/variables/src/provider.rs]: That's fair. I renamed the mod as part of 0a3e792

**Comment by ThorstenHans** [comment]
> Looks great overall! I did not actually run the code, but it looks pretty straight forward to me. I think we can merge this. The only points I had were nitpicks.

Awesome feedback @rylev I refactored the PR as part of 0a3e792

When implementing `FromStr`, I ended-up using `bail!` to fail early. Is that considered a valid practice here, or is there a better alternative?

File: crates/variables/src/provider/azure_key_vault.rs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+}
+impl FromStr for AzureAuthorityHost {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> std::prelude::v1::Result<Self, Self::Err> {
+        Ok(match s.to_lowercase().as_str() {
+            "azurechina" | "china" => AzureAuthorityHost::AzureChina,
+            "azuregermany" | "germany" => AzureAuthorityHost::AzureGermany,
+            "azuregovernment" | "gov" => AzureAuthorityHost::AzureGovernment,
+            "azureublic" | "azurepubliccloud" | "public" | "publiccloud" => {
```
Comment (rylev) [inline crates/variables/src/provider/azure_key_vault.rs:93]:
```suggestion
            "azurepublic" | "azurepubliccloud" | "public" | "publiccloud" => {
```

File: crates/variables/src/provider/azure_key_vault.rs [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        let url = match value {
+            AzureAuthorityHost::AzureChina => "https://login.chinacloudapi.cn/",
+            AzureAuthorityHost::AzureGovernment => "https://login.microsoftonline.us/",
+            AzureAuthorityHost::AzureGermany => "https://login.microsoftonline.de/",
+            AzureAuthorityHost::AzurePublicCloud => "https://login.microsoftonline.com/",
+        };
+        Url::parse(url).unwrap()
+    }
+}
+impl FromStr for AzureAuthorityHost {
```
Comment (rylev) [inline crates/variables/src/provider/azure_key_vault.rs:85]: I'm confused actually - what are the `FromStr` and `From<String>` impls for `AzureAuthorityHost` used for? It seems like they could just be deleted.

Comment (rylev) [reply crates/variables/src/provider/azure_key_vault.rs:85]: Hmm... `AzureKeyVaultProvider::new` only seems to be called in `AzureKeyVaultVariablesProviderOpts::build_provider` which passes a `AzureAuthorityHost` directly (not a string representation). That means that the only valid `AzureAuthorityHost` string representations in the runtime config file is whatever matches `AzureAuthorityHost`'s deserialize implementation which is just a one-to-one mapping with the Rust identifier. 

In other words, the runtime config will only accept the string "AzurePublicCloud" for the public cloud option. Other options like "public" and "publiccloud" won't work. Are you seeing something different in testing?

Comment (ThorstenHans) [reply crates/variables/src/provider/azure_key_vault.rs]: Makes sense. I tried without FromStr and `From<String>`. When users provide a wrong value for `AzureAuthorityHost` a corresponding error is presented and showing allowed values to the user - which is good. 

I rebased on most recent main and pushed the change once again

**Review by rylev — APPROVED** [review]
